packages/block-library/src/image/index.php

```diff
@@ -0,0 +1,357 @@
+	} else {
+		$image_dimensions = gutenberg_get_image_width_height( $block_attributes, $image_meta );
+	}
+
+	if ( empty( $image_dimensions ) || empty( $block_attributes['url'] ) ) {
+		// We don't have enough data to construct new img tag.
+		// Fall back to the existing HTML.
+		return $html;
+	}
+
+	$image_src = $block_attributes['url'];
+
+	$image_attributes = array(
+		'src' => $image_src,
+		'alt' => empty( $block_attributes['alt'] ) ? '' : $block_attributes['alt'],
+		'width' => $image_dimensions[0],
+		'height' => $image_dimensions[1],
+	);
+
+	if ( $image_meta ) {
+		// TODO: add another param so we can pass `$block_attributes` to the filter.
+		$srcset = wp_calculate_image_srcset( $image_dimensions, $image_src, $image_meta, $attachment_id );
+
+		if ( ! empty( $srcset ) ) {
+			// TODO: add another param so we can pass `$block_attributes` to the filter.
+			// This will let themes generate much better `sizes` attribute.
+			$sizes = wp_calculate_image_sizes( $image_dimensions, $image_src, $image_meta, $attachment_id );
+		}
+
+		if ( $srcset && $sizes ) {
+			$image_attributes['srcset'] = $srcset;
+			$image_attributes['sizes'] = $sizes;
+		}
+	}
+
+	/**
+	 * Filters the image tag attributes when rendering the core image block.
+	 *
+	 * @since 4.0
+	 *
+	 * @param array $image_attributes The (recalculated) image attributes.
+	 *                                Note: expects valid HTML 5.0 attribute names.
+	 * @param array $block_attributes The image block attributes.
+	 * @param string $html The image block HTML coming from the editor. The img tag will be replaced.
+	 */
+	$image_attributes = apply_filters( 'render_block_core_image_tag_attributes', $image_attributes, $block_attributes, $html );
+
+	$attr = '';
+	foreach ( $image_attributes as $name => $value ) {
+		// Sanitize for valid HTML 5.0 attribute names, see: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes.
+		$name = preg_replace( '/[^a-z0-9-]+/', '', strtolower( $name ) );
+
+		if ( empty( $name ) ) {
+			continue;
+		}
+
+		if ( 'src' === $name ) {
+			$value = esc_url( $value );
+		} elseif ( ( 'width' === $name || 'height' === $name ) && ! empty( $value ) ) {
+			$value = (int) $value;
+		} else {
+			$value = esc_attr( $value );
+		}
+
+		$attr .= sprintf( ' %s="%s"', $name, $value );
+	}
+
+	$image_tag = '<img' . $attr . '/ >';
+
+	// Replace the img tag.
+	$html = preg_replace( '/<img [^>]+>/', $image_tag, $html );
+
+	return $html;
+}
+
+function register_block_core_image() {
+	register_block_type(
+		'core/image',
+		array(
+			'attributes' => array(
+				'className' => array(
+					'type' => 'string', // Any custom class name(s) the user has entered.
+				),
+				'url' => array(
+					'type' => 'string', // The img src.
+				),
+				'alt' => array(
+					'type' => 'string',
+				),
+				'id' => array(
+					'type' => 'number',
+				),
+				'align' => array(
+					'type' => 'string',
+					'enum' => array( 'center', 'left', 'right', 'wide', 'full', '' ),
```

Comment:
From my testing in my own PRs, the `''` is unnecessary.

Reply:
Yeah, probably. However it makes that a bit more readable, i.e. all "possible values" are listed (so somebody looking at that place in couple of years won't think it's weird/buggy)  :) 

This was actually copied from https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/latest-comments/index.php#L177.

Reply:
Technically, `''` isn't actually a value that the attribute ever gets in practice. It technically gets `null` when no value is set, right? I mean, due to JavaScript type coercion, those 2 values end up being treated the same, but if you're going to put it there to make things clearer, I think putting `null` there would make more sense.

But by that logic, almost every instance of `enum` should include `null` or `''` when the attribute is optional for the block. I don't think it makes the code more readable; I think it actually createst confusion, because it implies that you can disallow an empty string or `null` value by removing it from the enum, which is not the case. Essentially, it's dead code that has no purpose, but implies that it does by existing.

I actually made #9832 to remove the `''` from the Latest Comments block a while back.